### PR TITLE
Improve loop performance in solvers

### PR DESCRIPTION
**Pull request by oscargus**
#### Brief description of what is fixed or changed

Use product and loop comprehensions.

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/solvers/ode/riccati.py [hunk 1]
```diff
@@ -663,9 +663,9 @@ def compute_m_ybar(x, poles, choice, N):
 
     # Calculate the first (nested) summation for ybar
     # as given in Step 9 of the Thesis (Pg 82)
-    for i in range(len(poles)):
-        for j in range(len(choice[i])):
-            ybar += choice[i][j]/(x - poles[i])**(j+1)
+    for i, pole in enumerate(poles):
+        for j, c in enumerate(choice[i]):
+            ybar += c/(x - pole)**(j+1)
```
Comment (oscarbenjamin) [inline sympy/solvers/ode/riccati.py:668]: If these are Expr then it is better to collect terms and sum the with `Add(*args)`. Iteratively adding in a loop is `O(n^2)` with Expr.

File: sympy/solvers/pde.py [hunk 2]
```diff
@@ -982,14 +982,8 @@ def _separate(eq, dep, others):
...
-        final = 0
-        for term in eq.args:
-            eqn = 0
-            for i in div:
-                eqn += term / i
-            final += simplify(eqn)
-        eq = final
-
+        # double sum required or some tests will fail
+        eq = sum(simplify(sum(term/i for i in div)) for term in eq.args)
```
Comment (oscarbenjamin) [inline sympy/solvers/pde.py:986]: This should use `Add(*...)` rather than `sum(...)`.

Comment (oscargus) [reply sympy/solvers/pde.py:986]: Yes, I realized that when seeing other similar code in my current rewrites.

File: sympy/solvers/ode/riccati.py (hunk 1 above)
Comment (smichr) [inline sympy/solvers/ode/riccati.py:668]:
```suggestion
    ybar = Add(*[c/(x - pole)**(j + 1) for i, pole in enumerate(poles)
        for j, c in enumerate(choice[i])])
```

File: sympy/solvers/ode/riccati.py [hunk 3]
```diff
@@ -663,9 +663,8 @@ def compute_m_ybar(x, poles, choice, N):
 
     # Calculate the first (nested) summation for ybar
     # as given in Step 9 of the Thesis (Pg 82)
-    for i in range(len(poles)):
-        for j in range(len(choice[i])):
-            ybar += choice[i][j]/(x - poles[i])**(j+1)
+    ybar = Add(*[c/(x - pole)**(j + 1) for i, pole in enumerate(poles)
+        for j, c in enumerate(choice[i])])
         m -= Poly(choice[i][0], x, extension=True)
```
Comment (smichr) [inline sympy/solvers/ode/riccati.py:668]:
```suggestion
    dybar = []
    dm = []
    for i, polei in enumerate(poles):
        for j, cij in enumerate(choice[i]):
            dybar.append(cij/(x - polei)**(j + 1)
        dm.append(Poly(choice[i][0], x, extension=True))
    ybar += Add(*dybar)
    m -= Add(*dm)
```

File: sympy/solvers/deutils.py [hunk 4]
```diff
@@ -121,15 +122,10 @@ def ode_order(expr, func):
...
-            for arg in expr.args[0].args:
-                order = max(order, ode_order(arg, func) + len(expr.variables))
-            return order
+            return max(ode_order(arg, func) for arg in expr.args[0].args) + len(expr.variables)
     else:
-        order = 0
-        for arg in expr.args:
-            order = max(order, ode_order(arg, func))
-        return order
+        return max(ode_order(arg, func) for arg in expr.args)
```
Comment (smichr) [inline sympy/solvers/deutils.py:127]:
```suggestion
        return max(ode_order(arg, func) for arg in expr.args, default=0)
```

File: sympy/solvers/deutils.py [hunk 5]
```diff
@@ -121,15 +122,10 @@ def ode_order(expr, func):
         if expr.args[0] == func:
             return len(expr.variables)
         else:
-            order = 0
-            for arg in expr.args[0].args:
-                order = max(order, ode_order(arg, func) + len(expr.variables))
-            return order
+            return max(ode_order(arg, func) for arg in expr.args[0].args) + len(expr.variables)
```
Comment (smichr) [inline sympy/solvers/deutils.py:125]:
```suggestion
            return max(ode_order(arg, func) for arg in expr.args[0].args, default=0
                ) + len(expr.variables)
```

File: sympy/solvers/deutils.py [hunk 6]
```diff
@@ -121,15 +122,11 @@ def ode_order(expr, func):
...
-                order = max(order, ode_order(arg, func) + len(expr.variables))
-            return order
+            return max(ode_order(arg, func) for arg in expr.args[0].args, default=0
+                ) + len(expr.variables)
     else:
-        order = 0
-        for arg in expr.args:
-            order = max(order, ode_order(arg, func))
-        return order
+        return max(ode_order(arg, func) for arg in expr.args, default=0)
```
Comment (smichr) [inline sympy/solvers/deutils.py:128]:
```suggestion
        return max(ode_order(_, func) for _ in expr.args) if expr.args else 0
```

File: sympy/solvers/deutils.py [hunk 7]
```diff
@@ -121,15 +122,14 @@ def ode_order(expr, func):
...
         else:
-            order = 0
-            for arg in expr.args[0].args:
-                order = max(order, ode_order(arg, func) + len(expr.variables))
-            return order
+            args = expr.args[0].args
+            rv = len(expr.variables)
+            if args:
+                rv += max(ode_order(_, func) for _ in args)
+             return rv
```
Comment (smichr) [inline sympy/solvers/deutils.py:129]:
```suggestion
            return rv
```

**Pull request merged**